### DISCUSSION TITLE
raw_turnstile.py not passing Cloudflare

### DISCUSSION THREAD

**Discussion by oculunto** [seleniumbase/SeleniumBase#2627] [General]
Hello, 
Just wanted to know if it was a general thing or just me, but it seems like Cloudflare is detecting clicks so the example raw_turnstile.py for example is not passing the check.
(It works properly with disconnect and clicking manually though).

**Comment by mdmintz** [comment]
Looks like Cloudflare made a change today.

Here's a Game of Thrones quote response:

![currently_at_work](https://github.com/seleniumbase/SeleniumBase/assets/6788579/4a185a24-58ee-4aed-981c-65d5781c3bab)

https://github.com/seleniumbase/SeleniumBase/assets/6788579/9481c7ed-995e-461a-a323-fda917f73879

**Reply by oculunto** [reply]
Hahaha thank you good sir ! 
I have tried to simulate mouse movement with actions but it gets me no where so far, I will keep trying.

**Comment by mdmintz** [comment]
As Qyburn would say about the newest version of CF's CAPTCHA checkbox:
It's "Powerful, but not invincible".

![powerful_but_not_invincible](https://github.com/seleniumbase/SeleniumBase/assets/6788579/8a9d03b1-4997-4343-8fe9-e1f5289a3735)

Now, "If you'll kindly pull that lever" after `pip install pyautogui` with the latest version of SeleniumBase:

```python
import pyautogui
from seleniumbase import SB

with SB(uc=True, test=True) as sb:
    sb.driver.uc_open_with_reconnect("https://seleniumbase.io/apps/turnstile")
    sb.send_keys("html", "\t")
    sb.driver.disconnect()
    pyautogui.press(" ")
    sb.sleep(1)
    sb.driver.connect()
    sb.assert_element("img#captcha-success", timeout=3)
    sb.set_messenger_theme(location="top_left")
    sb.post_message("SeleniumBase wasn't detected", duration=3)
```

<img width="800" alt="Screenshot 2024-03-20 at 11 00 44 PM" src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/f02e55c6-8d16-4df1-b8f9-43d8933d6c30">

Tested on macOS. You may have to enable permissions for the terminal to send keyboard actions.

The above solution currently requires knowing how many times to press the `Tab` key. Eg:

```python
sb.send_keys("html", "\t\t\t")
```

... before pressing the SPACEBAR with:

```python
pyautogui.press(" ")
```

This is only a temporary workaround. The plan is to make the click work without `pyautogui` by using pure `selenium`.

**Reply by oculunto** [reply]
Thank you for the workaround ! 
Unfortunately for me, I think the window needs to be in focus in order for pyautogui to click, and it doesn't work when having multiple chrome windows at the same time it seems...

**Reply by mdmintz** [reply]
That's why it's only a temporary workaround...
For people who couldn't wait a day for the next SeleniumBase release with a pure-Selenium solution.

**Reply by oculunto** [reply]
Yes I understand  
But it's a good idea, I'm working on a workaround based on it meanwhile.
Thanks for your work

**Reply by oculunto** [reply]
I am trying to use pywinauto instead and sending a virstual keystroke, but I cannot manage to get the PID of the window.
`browser.service.process.pid` does not retrieve a valid PID. 
Note that I am using many chrome processes at the same time, so it is not possible to just retrieve the PID by name.
Is there a way to get the current PID of the chrome window in SB?

**Reply by mdmintz** [reply]
Don't worry about it. The new release with the pure Selenium solution is coming later today.

**Comment by mdmintz** [comment]
Resolved in `4.24.12` - https://github.com/seleniumbase/SeleniumBase/releases/tag/v4.24.12

![good_scientist](https://github.com/seleniumbase/SeleniumBase/assets/6788579/a5ac28be-5055-4b97-bc96-01bcda4ab44f)

https://github.com/seleniumbase/SeleniumBase/assets/6788579/20e0703b-fdb7-43ec-9f69-2ad793c08895

----

Upgrade to the latest version, and try the updated examples. UC Mode is great again.

Notes: I examined the CF changes and found mistakes that they made in their latest release. The UC Mode method: `driver.uc_switch_to_frame(iframe)` isn't needed anymore because they aren't really looking for Selenium outside of their iframe (not including browser internals, such as console variables, which make a difference). This also means that you can fill out a form with Selenium, and then click the CAPTCHA checkbox (as opposed to before, where you had to click the checkbox first to minimize Selenium detection, before filling out a form). With the CF updates and the UC Mode updates to match their changes, UC Mode is now in better shape than before.

**Reply by oculunto** [reply]
![200w](https://github.com/seleniumbase/SeleniumBase/assets/111004839/bdb93e5d-ba03-46a0-87a3-3d868f64e810)
Thank you sir, boss !
